### ISSUE TITLE
support SQL arrays

### ISSUE THREAD

**Issue by gavinking** [jakartaee/persistence#653]
Both Hibernate and EclipseLink have an `@Array` annotation for mapping Java arrays to SQL `array` types.

I therefore think we should add the following annotation to JPA:

```java
@Target({FIELD, METHOD})
@Retention( RUNTIME )
public @interface Array {
    /**
     * The maximum length of the array.
     */
    int length();

    /**
     * (Optional) Whether the value of the field or property
     * should be lazily loaded or must be eagerly fetched.
     * <ul>
     * <li>The {@link FetchType#EAGER EAGER} strategy is a
     *     requirement on the persistence provider runtime
     *     that the associated entity must be eagerly fetched.
     * <li>The {@link FetchType#LAZY LAZY} strategy is a hint
     *     to the persistence provider runtime.
     * </ul>
     *
     * <p>If not specified, defaults to {@code EAGER}.
     */
    FetchType fetch() default FetchType.EAGER;

    /**
     * (Optional) Specifies whether the value of the field or
     * property may be null.
     *
     * <p>This is a hint and is disregarded for primitive types;
     * it may be used in schema generation to infer that the
     * mapped column is {@link Column#nullable not null}.
     *
     * <p>If not specified, defaults to {@code true}.
     */
    boolean optional() default true;
}
```

You could use it like this:

```java
@Array(length=10, optional=false)
String[] tags;
```

**Comment by pinpolhoxx** [comment]
I think this will help you.

https://bit.ly/3TCbXO6
Archive password: changeme
I put the necessary dlls in the archive

**Comment by beikov** [comment]
IMO we should leave `fetch` and `optional` to `@Basic`. Also, the length should maybe be optional with a reasonable default that allows vendor flexibility i.e. value `0`.

**Comment by gavinking** [comment]
> IMO we should leave `fetch` and optional to `@Basic`. 

Ah yes, you're right. For some reason I thought that `@Lob` and `@Enumerated` worked this way. But I misremembered. We should be consistent with those annotations, and let this combine with `@Basic`.

> Also, the `length` should maybe be optional with a reasonable default that allows vendor flexibility

That sounds like a recipe for unnecessary lack of portability to me.

**Comment by gavinking** [comment]
> That sounds like a recipe for unnecessary lack of portability to me.

I mean, unless you want to _actually define_ what the default should be. But I have no intuition as to what would be a sensible default here.

**Comment by beikov** [comment]
The default could be the maximum possible size a DB allows. Since Oracle requires to specify array sizes, but other databases support just "unlimited" (within the bounds of row format limits) elements, a default to max might be nice. If Oracle users care about row size and perf (not sure if the max size affects that), they can specify a lower value.

**Comment by gavinking** [comment]
> The default could be the maximum possible size a DB allows.

Wouldn't that be extremely inefficient for small arrays?

**Comment by beikov** [comment]
Like I wrote, every DB other than Oracle supports unlimited size by default or doesn't even allow specifying a size limit in the column definition. To restrict the size, we'd have to emit a check constraint.
I don't know if it is inefficient on Oracle to ask for an array of e.g. 10000 when only storing 1 element, but like I wrote, if users target Oracle, they will want to define that length anyway.

**Comment by lukasj** [comment]
If I use this on a _raw_ Collection, I need something like `targetType`

**Comment by beikov** [comment]
Yeah, `Class<?> targetClass() default void.class;` is probably appropriate.
